### ISSUE TITLE
dead seas b code

### ISSUE THREAD

**Issue by arkwise** [arkwise/deadseas_b#1]
since the last repo had issues @ArtemMurov we will upload the account here, i see that you closed your freelancer account? thats ok we will work from here.

@plongstaff will be working on the tcp side of the game, welcome to the team :)

**Comment by arkwise** [comment]
@plongstaff  whats up phil?

**Comment by plongstaff** [comment]
I am just getting my build environment set up with the 2020-09-20 source code. I assume your normal environment is a linux host cross-compiling using djgpp.

**Comment by arkwise** [comment]
well you can either cross compile through linux or used real dos or even compile under 32-bit windows in a DJGPP env

**Comment by plongstaff** [comment]
Do you have any specs yet for the server? 

I will transfer the freelancer questions here:

How do you envision that the server will run?
1) Large central server for everyone? You said "probably server nodes"
2) Large central server which will run multiple games and people can log into a specific one? You said "yes"
3) Server that people can download and run locally for a local game with friends? You said "yes"

The first thing to do, then, is to design the interaction between the game and the server.

**Comment by plongstaff** [comment]
Re development environment: yes, but it looks like your makefile and other scripts are currently set up for linux cross compile

**Comment by arkwise** [comment]
yes a central server with nodes for everyone that handles multiples games, chat and users that supports private network games in LAN

yes they are setup for linux cross compile because it uses a newer version of allegro, you could get this working under dos also too its up to you :)

**Comment by plongstaff** [comment]
I am starting to write server requirements doc.

[Server Requirements.txt](https://github.com/arkwise/deadseas_b/files/5298643/Server.Requirements.txt)

**Comment by plongstaff** [comment]
I have some thoughts about game communication.

Suppose you and I are playing. Each time I do something (buy a ship, upgrade a ship, move a ship, fire a cannon at your ship), an event message is sent to the server, and the server will forward to all attached players. Since we will use TCP/IP, events cannot be lost. However, events will still have a sequence number in case of disconnection. Thus, my game will know about everything you have done and your game will know about everything I have done. However, it may or may not be shown on the game map, depending on how you want to handle information. I assume I won't know about your ship movement unless I have a ship close enough to observe it.

How do you want to handle times when I am not logged in? Would you be able to approach and fire on my ships without me being able to respond, because I am not logged in? Can ships be set to defend themselves?

If I am looking at a certain part of the map, and in another part, you approach my ship close enough for me to see you, my game would receive the events, and could then pop up an inset showing that part of the map.

The server should accumulate event information so that it has the current state of the world: where each ship is, # hit points left, # ammo left, etc. When I attach to the game, the server should send the current state to my s/w. It should compare with my view of the world, and tell me what has changed (e.g. a ship has been sunk).

**Comment by plongstaff** [comment]
Once I have added more to the Server Requirements doc, I will add it to github. Text format is fine, though I can do it in Word if you prefer.

**Comment by plongstaff** [comment]
For deployment, I'm thinking of 2 servers:
1. Game server. This server would handle 1 game. This could run centrally or could be downloaded.
2. Dispatch server. There would be 1 of these and it would only run centrally. Would handle accounts and keep track of all of the game servers which are running centrally. If a new game is started, a new Game Server would be started which would listen on a new port for connections.

**Comment by arkwise** [comment]
> I have some thoughts about game communication.
> 
> Suppose you and I are playing. Each time I do something (buy a ship, upgrade a ship, move a ship, fire a cannon at your ship), an event message is sent to the server, and the server will forward to all attached players. Since we will use TCP/IP, events cannot be lost. However, events will still have a sequence number in case of disconnection. Thus, my game will know about everything you have done and your game will know about everything I have done. However, it may or may not be shown on the game map, depending on how you want to handle information. I assume I won't know about your ship movement unless I have a ship close enough to observe it.
> 
> How do you want to handle times when I am not logged in? Would you be able to approach and fire on my ships without me being able to respond, because I am not logged in? Can ships be set to defend themselves?
> 
> If I am looking at a certain part of the map, and in another part, you approach my ship close enough for me to see you, my game would receive the events, and could then pop up an inset showing that part of the map.
> 
> The server should accumulate event information so that it has the current state of the world: where each ship is, # hit points left, # ammo left, etc. When I attach to the game, the server should send the current state to my s/w. It should compare with my view of the world, and tell me what has changed (e.g. a ship has been sunk).

COOL,

basically you are correct, if the user completely drops out and times out they will just be simply removed from the game but the ships can stay there as a AI perhaps and defend themselves or be left as loot to be collected by other users.

the WIP DOC is coming along well, its basically what we needed before we lost a bunch of data earlier this year.

the dispatch server would be a good way for us to nut out the basics of the network play without all the log in info etc..

sorry for the late reply my computer finally died lol.

sooo just message me on freelancer or fb? it will get my attention faster.

really good work so far,

@ArtemMurov hey bud what are you working on at the moment did freelancer pay you the money yet??
